### PR TITLE
 Add more analytics graphs 

### PR DESCRIPTION
**Pull request by KevinBacabac**

### CODE REVIEW

File: discordwords.py [hunk 1]
```diff
@@ -10,9 +10,13 @@
...
 from pandas.api.types import CategoricalDtype
-import warnings
 
-import graphs.messages
+import resources.df_tools
+import resources.heatmap
+import resources.geoip
+import resources.messages
+import resources.reader
+import resources.tz
```
Comment (KevinBacabac) [inline discordwords.py:19]: New modules

File: discordwords.py [hunk 2]
```diff
@@ -23,264 +27,114 @@
 ps.add_argument("-s", "--start", type=str, nargs="+", help="Starting date (year/month/day) Defaults to beginning of data")
 ps.add_argument("-e", "--end", type=str, nargs="+", help="Stop date (year/month/day) Defaults to end of data")
 
-args=ps.parse_args()
+args = ps.parse_args()
```
Comment (KevinBacabac) [inline discordwords.py:30]: Automatic linting

File: discordwords.py [hunk 3]
```diff
@@ -23,264 +27,114 @@
...
-if len(channels)<1:
+
+if len(channels) < 1:
     print(args.path + " is not a readable discord data folder")
     exit()
 
-with open(os.path.join(args.path, "servers", "index.json")) as f:
-    servers = json.load(f)
-
-def load_cache(read_func, path, **kwargs):
```
Comment (KevinBacabac) [inline discordwords.py:42]: Move this block into separate modules

File: discordwords.py [hunk 4]
```diff
@@ -23,264 +27,114 @@
...
-        df[col] = pd.to_datetime(df[col])
-
-    # Localize and remove time but keep date
-    df[col] = df[col].dt.tz_localize(None).dt.normalize()
-
-    series = df[col].value_counts()
-    # Set missing dates to 0
-    series = series.resample('D').sum().sort_index()
-    return pd.DataFrame({col: series.index, 'Count': series.values})
+activity = resources.analytics.count_activity(args.path)
```
Comment (KevinBacabac) [inline discordwords.py:44]: Load data from modules

File: discordwords.py [hunk 5]
```diff
@@ -23,264 +27,114 @@
...
-for key, value in servers.items():
-    if key in fp_df["guild_id"]:
-        fp_df["guild_id"][value] = fp_df["guild_id"].pop(key)
-
-messages=[]
-acsv = pd.concat([load_cache(pd.read_csv, i, usecols=[1, 2]) for i in channels])
 acsv['Timestamp'] = pd.to_datetime(acsv['Timestamp'])
+acsv['Timestamp'] = resources.tz.localize_utc(acsv['Timestamp'])
 
-sdate=min(acsv['Timestamp'])
```
Comment (KevinBacabac) [inline discordwords.py:206]: Automatic linting

File: discordwords.py [hunk 6]
```diff
@@ -23,264 +27,114 @@
...
 if args.cloud:
     from wordcloud import WordCloud
     import matplotlib.pyplot as plt
-    print(str(str(len(twords))+' words selected over '+str(len(servers))+' servers.<br>'+
-           'Date range: '+str(pd.Timestamp(sdate).date())+' - '+str(pd.Timestamp(edate).date())+'\n'))
-    wordcloud = WordCloud(width=1920,height=1080, max_words=nmax,relative_scaling=1,normalize_plurals=False).generate_from_frequencies(dict(twords))
+    print(str(str(len(twords))+' words selected over '+str(len(servers))+' servers.<br>' +
+              'Date range: '+str(pd.Timestamp(sdate).date())+' - '+str(pd.Timestamp(edate).date())+'\n'))
+    wordcloud = WordCloud(width=1920, height=1080, max_words=nmax, relative_scaling=1,
+                          normalize_plurals=False).generate_from_frequencies(dict(twords))
```
Comment (KevinBacabac) [inline discordwords.py:112]: Automatic linting

File: discordwords.py [hunk 7]
```diff
@@ -23,264 +27,114 @@
...
     plt.axis("off")
     plt.show()
 
 
 else:
     fig = make_subplots(rows=4, cols=3, subplot_titles=("Total words", "Timeseries", "Messages per hour", "Messages per day",
-                                                        "Analytics", "City info", "IP info", "OS info",
-                                                        "Release channel", "Guild ID", "Event Type", "Private"))
+                                                        "Analytics", "Analytics Heatmap", "City info", "OS info",
+                                                        "Release channel", "Guild ID", "Event Type", "DMs vs Public"))
```
Comment (KevinBacabac) [inline discordwords.py:121]: Reorder and rename subplots

File: discordwords.py [hunk 8]
```diff
@@ -292,31 +146,26 @@ def get_stacked_area(key, value, **kwargs):
             mode='lines', stackgroup='one', groupnorm='percent',
             name=key, **kwargs)
 
-    for key, value in fp_df['city'].items():
-        fig.add_trace(get_stacked_area(key, value), row=2, col=3)
-
-    for key, value in fp_df['ip'].items():
-        fig.add_trace(get_stacked_area(key, value, showlegend=False), row=3, col=1)
-
-    for key, value in fp_df['os'].items():
```
Comment (KevinBacabac) [inline discordwords.py:301]: Combine into loop below

File: resources/df_tools.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    df.drop(PRUNE, axis=1, inplace=True)
+
+    df[col] = df[col].apply(parse_timestamp)
+
+    if interval == 'D':  # Remove time but keep date
+        df[col] = df[col].dt.normalize()
+
+    return count_df(df, col=col, interval=interval)
+
+def parse_timestamp(i):
```
Comment (KevinBacabac) [inline resources/df_tools.py:28]: We add support for old and new style timestamps

**Comment by wcshamblin** [comment]
Thanks for your contributions! Sorry for the delay, and I should be able to look over this soon - been very busy recently.

**Pull request merged**